src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +674,511 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+        }
+
+        co_return co_await do_run().finally([this] {
+            // Ensure all offset waiters have been removed before returning.
+            _as.request_abort();
+            return _waiter_gate.close();
+        });
+    }
+
+private:
+    shard_local_fetch_context _ctx;
+    ss::gate _waiter_gate;
+    ss::abort_source _as;
+
+    // This should roughly correspond to `_request_indexes.size()`. It's used to
+    // wait for partitions to have an offset change so that it's corresponding
+    // request in the fetch can be re-queried.
+    ssx::semaphore _completed_waiter_count{
+      0, "fetch_worker._completed_waiter_count"};
+    // This contains indexes into `_ctx.requests` and is populated by raft
+    // consensus waiters when an offset increases for a partition.
+    std::vector<size_t> _request_indexes;
+    // Contains the last observed value of `consensus->last_visible_index()` for
+    // every request in `_ctx.requests`. Its used to register waiters with
+    // `consensus->visible_offset_monitor()`.
+    std::vector<model::offset> _last_visible_indexes;
+
+    struct query_results {
+        std::vector<model::offset> last_visible_indexes;
+        // Indicates if any `read_result` in `results` has an error.
+        bool has_error;
+        std::vector<read_result> results;
+        size_t total_size;
+    };
+
+    ss::future<query_results>
+    query_requests(std::vector<ntp_fetch_config> requests) {
+        // The last visible indexes need to be populated before partitions
+        // are read. If they are populated afterwards then the
+        // last_visible_index could be updated after the partition is read,
+        // but before the index is queried. This leads to a race condition
+        // where we're waiting for the last_visible_index to change even
+        // though the partition has data we haven't read.
+        //
+        // Note that `last_visible_index` is used instead of the
+        // `commit_offset` as the commit offset won't change when partitions
+        // are produced to without acks=all. The `last_visible_index` more
+        // closely corresponds to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes(requests.size());
+        std::vector<std::tuple<size_t, model::partition_id>> errored_partitions;
+        size_t total_size{0};
+        bool has_error{false};
+
+        for (size_t i = 0; i < requests.size(); i++) {
+            const auto& req = requests[i];
+            auto part = _ctx.mgr.get(req.ktp());
+            if (!part) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            auto consensus = part->raft();
+            if (!consensus) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
```

Comment:
FWIW I feel like this is a little error prone. Mostly in this layer we're using kafka offsets (that have gone through translation), but here we explicitly are using raft offsets.

We should probably just prioritize using `kafka::offset` properly so this footgun is more explicit, but some documentation on this might help.

Reply:
Otherwise a quick look at this PR looks good. Nice work!